### DISCUSSION TITLE
LZ4 format / functionality extension to support HW implementations of LZ4

### DISCUSSION THREAD

**Discussion by quic-cgoldswo** [lz4/lz4#1517] [Ideas]
Hi Yann,

Qualcomm is designing a LZ4 compression / decompression engine, with the goal of being able to do single-pass operations (i.e. we only read input from DDR once for compression and decompression). This is achieved by using buffers internal to the engine that store:
- For compression, the running literal we've encountered, which is used as a search buffer 
- For decompression, the last part of the running decompressed output we've produced

The outcome of using internal (and obviously fixed-size) buffers for the above, whilst not making any changes to the LZ4 format, are as follows:
- For compression, if we fail to produce a match after running out of input buffer space, compression will fail. 
- For decompression, if the copy offset for a given block extends beyond what we're holding in our buffer, decompression will fail

Given this, we'd like to propose some optional extensions to LZ4 that would allow us to do the following:
1. Let our HW compress anything and let a SW implementation of LZ4 that supports our extensions to decompress the output.  
2. Let a SW implementation of LZ4 that supports our extensions compress any input, such that our HW will be able to subsequently decompress it 

For now, we're mainly interested in 1.  So, consider the following behavior change we would like to implement in our HW: if we're compressing with LZ4 and we hit our maximum literal size (i.e. run out of buffer space), **our idea is for our hardware to output a block containing the literal we have in our internal buffer, and an offset of 0, to indicate that we're outputting a literal without any subsequent copy**.  During decompression, if we encounter a block with an offset of 0, the literal contained in the block would be outputted, and we would continue to the next block without copying anything.

So, for HW-compression and SW-decompression, we'd like to propose that an offset of 0 be optionally interpreted as a command to dump the literal contained within a block, with no subsequent copy.  Since an offset of 0 is invalid and unused today, there should be no behavior breakage in supporting this.  For implementations that wouldn't support this extension and check for an offset of 0 as an error condition, decompression on compressed blocks from our extended LZ4 would fail outright. In reference to the Linux kernel's LZ4 library code, the change to the decompression code to handle such a block would be as follows:

https://elixir.bootlin.com/linux/v6.10.8/source/lib/lz4/lz4_decompress.c#L168
https://elixir.bootlin.com/linux/v6.10.8/source/lib/lz4/lz4_decompress.c#L292

```
diff -rupN "torvalds linux master lib-lz4/lz4_decompress.c" lib-lz4-patched/lz4_decompress.c
--- "torvalds linux master lib-lz4/lz4_decompress.c"  2024-08-11 09:51:42.000000000 -0700
+++ lib-lz4-patched/lz4_decompress.c  2024-08-12 06:26:33.986693000 -0700
@@ -166,6 +166,7 @@ static FORCE_INLINE int LZ4_decompress_g
      ip += 2;
      match = op - offset;
      assert(match <= op); /* check overflow */
+      if (unlikely(!offset)) continue; /* skip copy with zero offset */

      /* Do not deal with overlapping matches. */
      if ((length != ML_MASK) &&
@@ -289,6 +290,7 @@ static FORCE_INLINE int LZ4_decompress_g
    offset = LZ4_readLE16(ip);
    ip += 2;
    match = op - offset;
+    if (unlikely(!offset)) continue; /* skip copy with zero offset */

    /* get matchlength */
    length = token & ML_MASK;
```

We're aiming to merge a driver for the engine into Linux in the future, but we'd like to get your thoughts on this in advance.  Is it viable to add this as an optional feature to LZ4?
 
Thanks,
 
Chris Goldsworthy.

**Comment by Cyan4973** [comment]
This idea has been considered [for a few years already](https://github.com/lz4/lz4/issues/148#issuecomment-136810085),
and in the end, hasn't been implemented yet, out of a fear to introduce confusion into the LZ4 ecosystem.

What if people start to distribute LZ4-compressed frames and blocks, and it works on some systems, but not others ?
The entire existing LZ4 ecosystem would be impacted, as it would be tagged as unreliable with multiple similar stories cropping up.

Since then, I've been considering a few strategies to mitigate that risk.
Once again, I did not rush to implement them, since there was no urgency, but your topic might actually qualify as a pressuring event.

Now, how to mitigate risks of confusion ?
This could be a fairly elaborated multi-step process.

To begin with, we can make a clear distinction between LZ4 Frames and LZ4 Blocks.
Historically, LZ4 Blocks were developed first, with the view that this algorithm was designed as an application-specific library, so each application could decide how to transmit metadata (typically, compressed size, and an upper bound of decompressed size).
While this left room for customization, it also meant that many applications made different choices, making their implementations incompatible.
When appeared a need to share LZ4-compressed files between systems, problems started to crop up.
That's the reason why the LZ4 Frame format was created.

It means that, the LZ4 Frame Format has a clear mission of ensuring interoperability between implementations.
Whereas, the LZ4 Block format doesn't: different block-format implementations are already potentially incompatible.

This leaves room to introduce an updated LZ4 Block format, keeping it at block level initially, with the expectation that a Block implementation is only expected to be compatible with itself.
_Note_: it doesn't mean that this is a satisfying situation, it _should_ be improved. But at least it doesn't _break_ a current guarantee.

Now, introducing a new variant shouldn't be a hidden subtle event. Rather, it should be published, potentially discussed, as a way to prepare the ecosystem to the change over a long (multi-years) period. So this points towards updating the block format specification to `v2`.

The main change is, as proposed, to use the `offset` value `0` as a way to break long sections of literals, in order to ease buffer management for hardware implementations.
But since we are there, introducing a new variant, we might as well bundle a few other changes that have been put on ice for the same ecosystem compatibility reason.

The next change would be to drop the end-of-block special conditions. In hindsight, this was a relatively bad attempt to make the decoder easier to implement for correctness. But it bundled a few assumptions about how the decoder loop works, which are not invariant over time. Nowadays, the benefit of these special conditions are no longer present, but the nuisance on the encoding side are well present. So we might as well drop these special conditions, making the specification simpler.

Which lead to the next topic: what should be an end-of-block event ?
Of course, it's when input is exhausted, but that's not all. An input might be exhausted in the middle of something, and that would be invalid. We still have to define what's valid or not.
Previously, with the end-of-block conditions, the end-of-block itself was defined in a fairly rigid way: only at the end of a literals segment, which must be at least >= 5 literals.
Now, if we drop the end-of-block conditions, this rigidity is no longer warranted.
It could be at the of a literal segment, but the number of literals could be anything, 3, 2, 1, or even 0 bytes.
In this case, we might as well stop at the end of the previous match segment, no need to add a final `0` token.
Could it stop after a section splitter, aka a `00` offset ?
Well, why not? This could be a natural way to create long streams of `lz4` compressed data, with no prior knowledge of what follows (or even if there is anything to follow).
So that's quite a few more exit points from the loop.

What would still remain invalid is to stop in the middle of something, for example:
- only 1 byte of an 2-bytes offset provided
- literal segment incomplete (for example, it should contain 16 bytes, but only 15 are present)
- large length tokens unfinished (ending with a `0xFF`)

So, all in all, some fairly basic stuff, which may seem like secondary considerations, but are actually quite impactful for the code.

The good news is, after all these changes, a `V2` decoder remains 100% compatible with valid `V1` blocks.

There is possibly a discussion to have regarding invalid blocks: blocks which would be valid for `V2` would be considered successful by a `V2` decoder, while a `V1` would output an error. But if a `V1` block was expected, maybe the decoder is expected to output an error instead ? Sounds like a simple question, but once again, non-trivial consequences for the code.

And finally, the elephant in the room:
can these changes be applied without substantially degrading decoder performance?

Well, that's an unknown quantity at this point.
Last time I tried this change (many years ago), just introducing the offset `00` as a new way to "shortcut" the decoding loop by no longer looking for following match, decompression speed was affected by a significant -20%.
This new code path, I suspected, made the work on the compiler optimizer more difficult. One could imagine that some variables that were previously in registry might spill over to the stack ram instead.
I don't know for sure, but it shows that the outcome is not guaranteed a priori.

Of course, since it was many years ago, the outcome can be different today: compilers have improved, and may be more capable to deal with more complexity. Focus shifted from 32-bit to 64-bit, where we have a lot more registers. More importantly, the code is no longer the same (which is actually not a good thing: the current code is more efficient than the older one, but it's also significantly more complex, making it more difficult to update).

So, this entire effort, including ensuring an updated decoder would preserve the performance LZ4 is known for, this is just the first step. I understand that the first step could be enough for you, since it would unlock the possibility to generate special blocks with offset `00` from your hardware implementation and open the door to interoperability with software.

Eventually, interoperability with software is a more complex story, which requires a few more steps, some of which are more dangerous for the LZ4 ecosystem. 
But this is enough information for a single post already.

So maybe we could pause there and start a discussion.

**Reply by baybal** [reply]
So, you are making something akin ZSWAP in hardware, where the memory controller itself does the compression/decompression?

**Comment by quic-cgoldswo** [comment]
Hi Yann,

Though this only a single unit test, this is what I measured on an x86 build machine.  Note that I copied the tip of LZ4 into lzbench to collect these numbers, so the 1.9.3 version info below is incorrect.

Before:
```
$ ./lzbench -i100,100 -elz4 ../silesia/*
lzbench 1.8 (64-bit Linux)  Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
Assembled by P.Skibinski

Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                   5301 MB/s  5295 MB/s    10192446 100.00 ../silesia/dickens
lz4 1.9.3                 276 MB/s  2652 MB/s     6428742  63.07 ../silesia/dickens
lz4 1.9.3                 456 MB/s  2592 MB/s    26435667  51.61 ../silesia/mozilla
lz4 1.9.3                 441 MB/s  3010 MB/s     5440937  54.57 ../silesia/mr
lz4 1.9.3                 766 MB/s  3612 MB/s     5533040  16.49 ../silesia/nci
lz4 1.9.3                 367 MB/s  2417 MB/s     4338918  70.53 ../silesia/ooffice
lz4 1.9.3                 389 MB/s  2400 MB/s     5256666  52.12 ../silesia/osdb
lz4 1.9.3                 287 MB/s  2345 MB/s     3181387  48.00 ../silesia/reymont
lz4 1.9.3                 480 MB/s  2931 MB/s     7716839  35.72 ../silesia/samba
lz4 1.9.3                 365 MB/s  3169 MB/s     6790273  93.63 ../silesia/sao
lz4 1.9.3                 319 MB/s  2459 MB/s    20139988  48.58 ../silesia/webster
lz4 1.9.3                 618 MB/s  2950 MB/s     1227495  22.96 ../silesia/xml
lz4 1.9.3                 983 MB/s  6005 MB/s     8390195  99.01 ../silesia/x-ray
done... (cIters=100 dIters=100 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
```

After:
```
$ ./lzbench -i100,100 -elz4 ../silesia/*
lzbench 1.8 (64-bit Linux)  Intel(R) Xeon(R) Silver 4114 CPU @ 2.20GHz
Assembled by P.Skibinski

Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                   5291 MB/s  5301 MB/s    10192446 100.00 ../silesia/dickens
lz4 1.9.3                 277 MB/s  2666 MB/s     6428742  63.07 ../silesia/dickens
lz4 1.9.3                 457 MB/s  2598 MB/s    26435667  51.61 ../silesia/mozilla
lz4 1.9.3                 441 MB/s  3013 MB/s     5440937  54.57 ../silesia/mr
lz4 1.9.3                 767 MB/s  3618 MB/s     5533040  16.49 ../silesia/nci
lz4 1.9.3                 367 MB/s  2437 MB/s     4338918  70.53 ../silesia/ooffice
lz4 1.9.3                 389 MB/s  2395 MB/s     5256666  52.12 ../silesia/osdb
lz4 1.9.3                 287 MB/s  2345 MB/s     3181387  48.00 ../silesia/reymont
lz4 1.9.3                 481 MB/s  2937 MB/s     7716839  35.72 ../silesia/samba
lz4 1.9.3                 365 MB/s  3195 MB/s     6790273  93.63 ../silesia/sao
lz4 1.9.3                 320 MB/s  2463 MB/s    20139988  48.58 ../silesia/webster
lz4 1.9.3                 619 MB/s  2952 MB/s     1227495  22.96 ../silesia/xml
lz4 1.9.3                 984 MB/s  6031 MB/s     8390195  99.01 ../silesia/x-ray
done... (cIters=100 dIters=100 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
```

So, the results seem to be unchanged, though this is a single data point.  In the past, did you take any steps to try testing things at scale on different architectures / CPUs?  On my end, I only have access to phone SoCs and build workstations. 

Thanks,

Chris.

P.S. In the Linux kernel, the core kernel / kenrel module binaries are modifiable at run time, in order to support features like [static keys for branch elimination when some feature is not present](https://www.kernel.org/doc/Documentation/static-keys.txt).  So we would probably end up opting for that on our end.  But that obviously doesn't generalize.
